autogpt_platform/backend/backend/integrations/oauth/google.py

```diff
@@ -35,36 +46,78 @@ def get_login_url(self, scopes: list[str], state: str) -> str:
         flow.redirect_uri = self.redirect_uri
-        flow.fetch_token(code=code)
 
-        google_creds = flow.credentials
-        username = self._request_email(google_creds)
+        logger.debug("Fetching token from Google")
 
-        # Google's OAuth library is poorly typed so we need some of these:
-        assert google_creds.token
-        assert google_creds.refresh_token
-        assert google_creds.expiry
-        assert google_creds.scopes
-        return OAuth2Credentials(
-            provider=self.PROVIDER_NAME,
-            title=None,
-            username=username,
-            access_token=SecretStr(google_creds.token),
-            refresh_token=SecretStr(google_creds.refresh_token),
-            access_token_expires_at=int(google_creds.expiry.timestamp()),
-            refresh_token_expires_at=None,
-            scopes=google_creds.scopes,
-        )
+        try:
+            # Disable scope check in fetch_token
+            flow.oauth2session.scope = None
+            token = flow.fetch_token(code=code)
+            logger.debug("Token fetched successfully")
+
+            # Get the actual scopes granted by Google
+            granted_scopes: list[str] = s.split() if (s := token.get("scope")) else []
+
+            logger.debug(f"Scopes granted by Google: {granted_scopes}")
+
+            google_creds = flow.credentials
+            logger.debug(f"Received credentials: {google_creds}")
+
+            logger.debug("Requesting user email")
+            username = self._request_email(google_creds)
+            logger.debug(f"User email retrieved: {username}")
+
+            if not google_creds.token:
+                raise ValueError("No access token received")
+            if not google_creds.refresh_token:
+                raise ValueError("No refresh token received")
+            if not google_creds.expiry:
+                raise ValueError("No expiry time received")
+            if not granted_scopes:
+                raise ValueError("No scopes received")
+
+            # Create OAuth2Credentials with the granted scopes
+            credentials = OAuth2Credentials(
+                provider=self.PROVIDER_NAME,
+                title=None,
+                username=username,
+                access_token=SecretStr(google_creds.token),
+                refresh_token=(SecretStr(google_creds.refresh_token)),
+                access_token_expires_at=(
+                    int(google_creds.expiry.timestamp())
+                    if google_creds.expiry
+                    else None
+                ),
+                refresh_token_expires_at=None,
+                scopes=granted_scopes,
+            )
+            logger.debug(
+                f"OAuth2Credentials object created successfully with scopes: {credentials.scopes}"
+            )
+
+            return credentials
+
+        except Exception as e:
+            logger.error(f"Error during code exchange: {str(e)}", exc_info=True)
+            raise
```

Comment:
Still double-stacking the error handling. This is redundant with the try/except in the API endpoint, which has a more specific/verbose log statement.

Reply:
removed